File: configs/client/python/producer.py
from confluent_kafka import Producer
import logging
import socket
import time

logging.basicConfig(level=logging.INFO, format='%(asctime)s %(levelname)s %(message)s')
logger = logging.getLogger(__name__)

config = {
    'bootstrap.servers': 'broker:9092',
    'security.protocol': 'SASL_SSL',
    'sasl.mechanism': 'SCRAM-SHA-256',
    'sasl.username': 'cagri',
    'sasl.password': 'cagri3541',
    'ssl.ca.location': '/mnt/home/python/client.pem',
    'ssl.endpoint.identification.algorithm': 'none',
    'client.id': f'python-producer-{socket.gethostname()}',
    'socket.timeout.ms': 30000,
    'retry.backoff.ms': 500
}

TOPIC_NAME = 'cagri-a'

def delivery_callback(err, msg):
    """Callback function for message delivery reports"""
    if err is not None:
        logger.error(f"Failed to send message: {err}")
    else:
        msg_count = int(msg.value().decode('utf-8').replace('cagri', ''))
        if msg_count % 10000 == 0:
            logger.info(f"Sent {msg_count} messages, offset: {msg.offset()}")

try:
    producer = Producer(config)
    logger.info(f"Created producer, sending messages to {TOPIC_NAME}")

    for i in range(1, 100000001):
        message = f"cagri{i}"
        producer.produce(
            topic=TOPIC_NAME,
            value=message.encode('utf-8'),
            callback=delivery_callback
        )

        producer.poll(0)

        if i % 10000 == 0:
            producer.flush()
            
except Exception as e:
    logger.error(f"Error sending messages: {e}")
    import traceback
    traceback.print_exc()
finally:
    logger.info("Flushing and closing producer")
    producer.flush()

File: configs/client/scala/scala_files/producer.scala
import org.apache.kafka.clients.producer.{KafkaProducer, ProducerConfig, ProducerRecord}
import org.apache.kafka.common.serialization.StringSerializer
import java.util.Properties

object Producer extends App {
  System.setProperty("java.security.krb5.conf", "/etc/krb5.conf")
  System.setProperty("java.security.auth.login.config", "/mnt/home/scala/client_jaas.conf")

  val props = new Properties()
  props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "broker:9092")
  props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, classOf[StringSerializer].getName)
  props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, classOf[StringSerializer].getName)
  
  // SASL configs
  props.put("security.protocol", "SASL_SSL")
  props.put("sasl.mechanism", "GSSAPI")
  props.put("sasl.kerberos.service.name", "kafka")
  props.put("ssl.truststore.location", "/mnt/home/scala/client.truststore.jks")
  props.put("ssl.truststore.password", "cagri3541")
  props.put("ssl.truststore.type", "PKCS12")
  props.put("ssl.endpoint.identification.algorithm", "")

  val producer = new KafkaProducer[String, String](props)
  val TOPIC_NAME = "cagri-a"

  try {
    for (i <- 1 to 1000000) {
      val record = new ProducerRecord[String, String](TOPIC_NAME, s"cagri$i")
      producer.send(record, (metadata, exception) => {
        if (exception != null) {
          println(s"Failed to send message: ${exception.getMessage}")
        } else if (i % 10000 == 0) {
          println(s"Sent $i messages, offset: ${metadata.offset()}")
        }
      })
      if (i % 10000 == 0) producer.flush()
    }
  } catch {
    case e: Exception =>
      println(s"Error sending messages: ${e.getMessage}")
      e.printStackTrace()
  } finally {
    producer.close()
  }
}